qiskit/transpiler/target.py

```diff
@@ -0,0 +1,600 @@
+                instruction name in ``inst_map`` to an instruction object
+            error_dict (dict): A dictionary of errors of the form::
+
+                {gate_name: {qarg: error}}
+
+            for example::
+
+                {'rx': {(0, ): 1.4e-4, (1, ): 1.2e-4}}
+
+            For each entry in the ``inst_map`` if ``error_dict`` is defined
+            a when updating the ``Target`` the error value will be pulled from
+            this dictionary. If one is not found in ``error_dict`` then
+            ``None`` will be used.
+
+        Raises:
+            ValueError: If ``inst_map`` contains new instructions and
+                ``inst_name_map`` isn't specified
+            KeyError: If a ``inst_map`` contains a qarg for an instruction
+                that's not in the target
+        """
+        for inst in inst_map.instructions:
+            out_props = {}
+            for qarg in inst_map.qubits_with_instruction(inst):
+                sched = inst_map.get(inst, qarg)
+                val = InstructionProperties(schedule=sched)
+                try:
+                    qarg = tuple(qarg)
+                except TypeError:
+                    qarg = (qarg,)
+                if inst in self._gate_map:
+                    if self.dt is not None:
+                        val.length = sched.duration * self.dt
+                    else:
+                        val.length = None
+                    if error_dict is not None:
+                        error_inst = error_dict.get(inst)
+                        if error_inst:
+                            error = error_inst.get(qarg)
+                            val.error = error
+                        else:
+                            val.error = None
+                    else:
+                        val.error = None
+                out_props[qarg] = val
+            if inst not in self._gate_map:
+                if inst_name_map is not None:
+                    self.add_instruction(inst_name_map[inst], out_props, name=inst)
+                else:
+                    raise ValueError(
+                        "An inst_name_map kwarg must be specified to add new "
+                        "instructions from an InstructionScheduleMap"
+                    )
+            else:
+                for qarg, prop in out_props.items():
+                    self.update_instruction_properties(inst, qarg, prop)
+
+    @property
+    def qargs(self):
+        """The set of qargs in the target."""
+        if None in self._qarg_gate_map:
+            return None
+        return set(self._qarg_gate_map)
+
+    def get_qargs(self, gate):
+        """Get the qargs for a given gate name
+
+        Args:
+           gate (str): The gate instance to get qargs for
+        Returns:
+            set: The set of qargs the gate instance applies to
+        """
+        if None in self._gate_map:
+            return None
+        return set(self._gate_map[gate])
+
+    def durations(self):
```

Comment:
how do you choose whether something is a `property` or `method`?

Reply:
Mostly I was trying to make things that return separate objects methods especially those for existing transpiler data objects, like `CouplingMap` (coupling map above was an oversight from a much earlier draft and should be a method), `InstructionDuration`, `InstructionScheduleMap`. While things that return lists or objects which use basic containers (like lists or dicts) that a user might want to directly access are properties (so you can do something like `target.instructions[2]`). But I might not have applied things consistently especially as this PR has been in progress for a very long time and is quite large I might have missed things (like I did for coupling map)